### PR TITLE
NMS-13268: Karaf: Globally Prevent AutoRefresh Cascade on Feature Install

### PR DESCRIPTION
**Pull request by swachter**
Issue: https://issues.opennms.org/browse/NMS-13268

This PR configures Karaf such that bundles are not refreshed when features are installed. Without that change a bundle is stopped / started if a newly installed feature brings an optional dependency or a newer version of a dependency of that bundle. That seems to be unnecessary in case of OpenNMS Karaf setups.

In addition, this PR reformats two lines of code where the flaky `BMSTopologyIT.canChangeFocusWhenInSimulationMode` test randomly fails with a NPE. Next time that test failes, more precise NPE information may help to fix the flakiness of that test.

### CODE REVIEW

**Review by fooker — APPROVED** [review]

**Comment by j-white** [comment]
May be okay, but I can't say that I'm confident this won't have some unintended side effects.

How much does this impact startup time?

**Comment by RangerRick** [comment]
> May be okay, but I can't say that I'm confident this won't have some unintended side effects.

My fear as well, but I guess we have an entire release schedule to make sure it works OK... :D

**Review by RangerRick — DISMISSED** [review]

**Comment by swachter** [comment]
> May be okay, but I can't say that I'm confident this won't have some unintended side effects.
> 
> How much does this impact startup time?

I attached some log outputs and added a comment with some numbers to the ticket: [NMS-13268](https://issues.opennms.org/browse/NMS-13268?focusedCommentId=66317&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-66317).

TLDR: For the Alec plugin the speedup was 669 ms.

Agreeable, that may not be worth the change. OTOH having less bundles going up and down may be a benefit on its own. I also do not have experience on the number of plugins in typical OpenNMS installations.

One by-product of my investigation was that the way features are installed by the `KarafExtender` is not influenced by the `autoRefresh` configuration setting. AutoRefresh behavior of the `KarafExtender` must be set by giving a parameter when calling the `FeatureService`.

Another by-product was that it may be worth to investigate what is happening between feature installs. Startup pauses in the range of 10s of seconds.

**Comment by swachter** [comment]
Some more information: Karaf 4.3.2 has a bug that prevents the usage of the `autoFlush` configuration setting (cf. [KARAF-7282](https://issues.apache.org/jira/browse/KARAF-7282)). Auto refresh can be prevented programmatically only.

**Comment by RangerRick** [comment]
Sounds like we should revisit this when 4.3.4 is out.

**Comment by swachter** [comment]
> Sounds like we should revisit this when 4.3.4 is out.

If we wait for 4.3.4 then auto-refresh on feature installs can be completely disabled by the global configuration setting `autoRefresh=false`. This will apply to both, feature installs on startup and manual feature installs. No code changes are necessary to `KarafExtender` and `KarDeployer`.

If we do not wait for 4.3.4 then we could disable auto-refreshs on startup only.

In summary, I also opt for waiting until we upgrade to Karaf 4.3.4.

**Comment by RangerRick** [comment]
We need to update to 4.3.4 to continue to work on JDK17 support anyway, so it's gonna be a priority going forward.

**Comment by RangerRick** [comment]
@swachter should be able to look at this again as soon as https://github.com/OpenNMS/opennms/pull/4112 is merged

**Comment by RangerRick** [comment]
Hrm, the failures in `smoke-test-flaky` look like they could actually be from things not loading right...?

**Comment by swachter** [comment]
> Hrm, the failures in `smoke-test-flaky` look like they could actually be from things not loading right...?

It is each time a different one: `ReportRestIT`, `canChangeFocusWhenInSimulationMode`, and `verifyBmpProcessing`. Not sure what that means, either. I am inclined that the change does not introduce a problem. I will have a closer look on the logs, though.

**Comment by RangerRick** [comment]
> > Hrm, the failures in `smoke-test-flaky` look like they could actually be from things not loading right...?
> 
> It is each time a different one: `ReportRestIT`, `canChangeFocusWhenInSimulationMode`, and `verifyBmpProcessing`. Not sure what that means, either. I am inclined that the change does not introduce a problem. I will have a closer look on the logs, though.

I think it is unlikely it's a pre-existing condition since I don't see these kinds of errors in the main build. We _do_ get some flapping, but not in this particular way. This almost looks like some kind of race condition or missing dependency or something.

**Comment by swachter** [comment]
* The failure in `ReportRestIT` relates to [NMS-13961](https://issues.opennms.org/browse/NMS-13961) which is hopefully fixed soon.

* The failure in `canChangeFocusWhenInSimulationMode` is caused by a NPE in `GraphPainter.getSourceKey(Edge)`. In principle, that method is allowed to return `null` but does no `null` check if the vertex exists before `.getKey()` is called:

```
	private String getSourceKey(Edge edge) {
		return m_graphContainer
				.getTopologyServiceClient()
				.getVertex(
						edge
								.getSource()
								.getVertex(),
						m_graphContainer.getCriteria()
				)
				.getKey();
	}
```

Admittedly, this seems to be a timing issue because sometimes that test fails and sometimes it does not. This PR changes the startup timing. However, I think that the issue should be resolved on its own. I am tempted to "fix" the issue by just adding a null check for the vertex. Maybe someone has more insight into this??? WDYT?

* Unfortunately, I have no idea about the failed `verifyBmpProcessing` test.

**Comment by RangerRick** [comment]
> * In principle, that method is allowed to return `null` but does no `null` check if the vertex exists before `.getKey()` is called

Yeah, this has been flapping elsewhere as well, not just your branch. Does seem correct to null-check, and it affects all the way back to `foundation-2019` so I'll play with it there.

**Review by RangerRick — DISMISSED** [review]
assuming it passes full smoke once you fix conflicts/merge the updates, LGTM

**Comment by cgorantla** [comment]
> > assuming it passes full smoke once you fix conflicts/merge the updates, LGTM
> 
> There were always some smoke tests failing even after 3 "return from failed":
> 
> * rerun: `SyslogKafkaElasticsearchIT`
> * rerun: `testHeartbeatOutages`
> * initial run: `verifyStatusEnrichmentApplication`, `testSnmpV3TrapsOnMinion`, and `testRestHealthServiceOnSentinel`
> 
> Either I am extremely unlucky or this change makes our tests indeed more flaky. I am unsure about this PR but tend to just close it.

Some of these are flaky in general.
Overall, Since we are not sure what are the side effects of this change, i would vote for closing it for now.
We can revisit this later.
